File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.js
function solution(triangle) {
  const height = triangle.length;

  for (let row = 1; row < height; row++) {
    const len = triangle[row].length;
    for (let col = 0; col < len; col++) {
      let val;
      if (col === 0) {
        val = triangle[row][col] + triangle[row - 1][col];
      } else if (col === len - 1) {
        val = triangle[row][col] + triangle[row - 1][col - 1];
      } else {
        val =
          triangle[row][col] +
          Math.max(triangle[row - 1][col - 1], triangle[row - 1][col]);
      }
      triangle[row][col] = val;
    }
  }
  // console.log(triangle[height - 1]);

  return Math.max(...triangle[height - 1]);
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.py
def solution(triangle):
  N = len(triangle)
  K = len(triangle[N-1])
  
  board = [[0] * (K+1) for _ in range(N)]
  for row in range(N):
    for col,value in enumerate(triangle[row]):
      # 첫번째 줄
      if(row == 0):
        board[row][col+1] = value
      else:
        board[row][col+1] = value + max(board[row-1][col+1], board[row-1][col])
  
  # 최대값
  return max(board[N-1])